### PR TITLE
components: Add new ColorPicker

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds a new vastly simplified ColorPicker implementation based on `react-colorful`.

## How has this been tested?
Storybook.

## Screenshots 

https://user-images.githubusercontent.com/24264157/127386106-606631ef-2d8e-4c5a-973f-c288bddb947d.mov

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/tsconfig.json [hunk 1]
```diff
@@ -34,17 +35,23 @@
 		"src/grid/**/*",
 		"src/h-stack/**/*",
 		"src/heading/**/*",
+		"src/input-control/**/*",
 		"src/icon/**/*",
-		"src/item-group/**/*",
```
Comment (ciampo) [inline packages/components/tsconfig.json:39]: This line was recently added in #33701, I don't believe it should be deleted

File: packages/components/src/number-control/styles/number-control-styles.js [hunk 2]
```diff
@@ -17,6 +18,10 @@ const htmlArrowStyles = ( { hideHTMLArrows } ) => {
 			-webkit-appearance: none !important;
 			margin: 0 !important;
 		}
+
+		input[type='number'] {
+			-moz-appearance: textfield;
+		}
```
Comment (ciampo) [inline packages/components/src/number-control/styles/number-control-styles.js:24]: What is this snippet achieving?

File: packages/components/src/range-control/index.js [hunk 3]
```diff
@@ -194,6 +196,7 @@ function RangeControl(
 		<BaseControl
 			className={ classes }
 			label={ label }
+			hideLabelFromVision={ hideLabelFromVision }
```
Comment (ciampo) [inline packages/components/src/range-control/index.js:199]: Why was this prop added?

File: packages/components/src/select-control/index.tsx [hunk 4]
```diff
@@ -18,13 +20,38 @@ import BaseControl from '../base-control';
...
+	options?: {
+		label: string;
+		value: string;
+		id?: string;
+		disabled?: boolean;
+	}[];
+	size?: 'small' | 'default';
+	value?: string | string[];
+	labelPosition?: 'top' | 'side' | 'bottom';
+}
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx]: Are all of these props supposed to be optional?

File: packages/components/src/ui/color-picker/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+**Type**: `(hex8Color: string) => void`
+
+Fired when the color changes. Always passes a hex8 color string.
+
+### `enableAlpha`
+
+**Type**: `boolean`
+
+Defaults to `false`. When `true` the color picker will display the alpha channel both in the bottom inputs as well as in the color picker itself.
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/README.md:45]: Nit: should we use the same format we've been using for newer components?

```suggestion
### `enableAlpha`: `boolean`

 When `true` the color picker will display the alpha channel both in the bottom inputs as well as in the color picker itself.
 
- Required: No
- Default: `false`
```

File: packages/compose/src/hooks/use-copy-to-clipboard/index.js [hunk 6]
```diff
@@ -27,11 +27,12 @@ function useUpdatedRef( value ) {
 /**
  * Copies the given text to the clipboard when the element is clicked.
  *
+ * @template {HTMLElement} TElementType
  * @param {string | (() => string)} text      The text to copy. Use a function if not
  *                                            already available and expensive to compute.
  * @param {Function}                onSuccess Called when to text is copied.
  *
- * @return {import('react').Ref<HTMLElement>} A ref to assign to the target element.
+ * @return {import('react').Ref<TElementType>} A ref to assign to the target element.
```
Comment (ciampo) [inline packages/compose/src/hooks/use-copy-to-clipboard/index.js:35]: Why was this change necessary?

File: packages/components/src/ui/color-picker/rgba-color-picker.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentProps } from 'react';
+import { RgbaColorPicker as Picker } from 'react-colorful';
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/rgba-color-picker.tsx]: Nit: when renaming imports from `react-colorful`, should we actually prefix these components with `ReactColorful*` or something similar?

File: packages/components/src/ui/color-picker/input-with-slider.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	flex: 1;
+
+	${ StyledField } {
+		margin-bottom: 0;
+	}
+`;
+
+const Wrapper = styled( HStack )`
+	margin-bottom: ${ space( 2 ) };
+`;
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/input-with-slider.tsx]: Should these be moved to the `styles` file and be suffixed with `*View` ?

File: packages/components/src/ui/color-picker/styles.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ */
+import InnerSelectControl from '../../select-control';
+import { StyledField } from '../../base-control/styles/base-control-styles';
+import { space } from '../utils/space';
+
+export const SelectControl = styled( InnerSelectControl )`
+	width: 5em;
+`;
+
+export const ColorfulWrapper = styled.div`
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/styles.ts]: Should we add the `*View` suffix to the components in this file?

**Review by ciampo — COMMENTED** [review]
Nicely done so far, @sarayourfriend! This was quite a large PR to get through.

I've left a few inline comments, but I also have a few more general pieces of feedback:

- I haven't reviewed any `InputControl`-related changes because I assume that they're actually being worked on in a separate PR #33696 , and that this PR will be rebased once those changes land on `trunk`
- Is there a reason why the conversion to TS of the `SelectControl` component isn't also done in a separate PR?
- Should we make some small changes to the APIs (props) of the `ColorPicker` component, in order to match the ["legacy" `ColorPicker`](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/color-picker)?
- I haven't done an exhaustive review in terms of accessibility, but at a quick glance I can see that focus styles are not very prominent (not sure if the original designs have any specs for that) cc @diegohaz 

Comparing the Storybook prototype against the latest designs, I could spot a few differences, which I annotated over a couple of screenshots:

image

image

image

**Comment by sarayourfriend** [comment]
Yup, definitely aware of the design differences 🙂 

However, I'm not sure how best to address them.

On one hand, we can change the default styles for these components to match the designs. This would update all of Gutenberg.

On the other hand, we could create wrappers for all these components and manually change the styles using `styled` components and careful targeting. This would have two problems in my book:

1. It would present one section of the UI as completely visually separate from the rest of the UI
2. It would replicate the complexities of Sass based styling, i.e., targeting and overriding specific classes on the components we want to manipulate

It would also be code we would hopefully anticipate deleting in the near future when the designs of the underlying components were updated.

My preference is for the first approach, simply update the existing component designs.

However, of course, I don't want to do that all in one PR. So in some sense this PR isn't at all mergable not because it's incomplete but becuase it's uncovered a dearth of dependencies that need to be resolved before it can be merged (esentially all of the visual issues you pointed out save the spacing ones).

File: packages/compose/src/hooks/use-copy-to-clipboard/index.js (hunk 6 above)
Comment (sarayourfriend) [reply packages/compose/src/hooks/use-copy-to-clipboard/index.js:35]: Because `Ref<HTMLElement>` isn't, assignable to `Ref<HTMLDivElement>` (but the other way around is allowed)

File: packages/components/src/range-control/index.js (hunk 3 above)
Comment (sarayourfriend) [reply packages/components/src/range-control/index.js:199]: It's a prop that all components that extends BaseControl expose from BaseControl. I'd see this as a bug in the API of the RangeControl component.

File: packages/components/src/number-control/styles/number-control-styles.js (hunk 2 above)
Comment (sarayourfriend) [reply packages/components/src/number-control/styles/number-control-styles.js:24]: Because the existing code didn't work for Firefox to hide the HTML arrows in number inputs.

**Comment by sarayourfriend** [comment]
> Should we make some small changes to the APIs (props) of the ColorPicker component, in order to match the "legacy" ColorPicker?

I don't think we should take _any_ cues from the existing ColorPicker component. It's massively over complicated.

File: packages/components/src/range-control/index.js (hunk 3 above)
Comment (ciampo) [reply packages/components/src/range-control/index.js:199]: Cool! Do you think it's worth mentioning in the changelog as a fix?

File: packages/components/src/number-control/styles/number-control-styles.js (hunk 2 above)
Comment (ciampo) [reply packages/components/src/number-control/styles/number-control-styles.js:24]: Makes sense, thank you for the explanation!

**Comment by ciampo** [comment]
> I don't think we should take _any_ cues from the existing ColorPicker component. It's massively over complicated.

If we wanted this `ColorPicker` to be a drop-in replacement for the existing `ColorPicker`, shouldn't we make sure that the new one has the same props as the legacy one (or that, at least, there is a `useDeprecatedProps` compat-layer which is able to parse the props of the legacy `ColorPicker` and adapt them to the new one)?

I know that this can be done at a later point in time when we promote the g2 `ColorPicker`, but I thought it'd be better to discuss it sooner rather than later.

File: packages/components/src/range-control/index.js (hunk 3 above)
Comment (ciampo) [reply packages/components/src/range-control/index.js:199]: Done in [`dc448d4` (#33714)](https://github.com/WordPress/gutenberg/pull/33714/commits/dc448d4ce86ae1ec95bfbefd4dd8e1904a58f2d5)

**Comment by sarayourfriend** [comment]
I think my proposal for the ColorPicker is to add a prop like `__unstableSimplified` which switches between the legacy and the new color picker, instead of trying to adapt the legacy props.

Primarily the issue I have the with the existing ColorPicker is that it's API is far too broad, especially in the change handlers.

It also uses "backwards" props like `disableAlpha` instead of positive boolean props like `enableAlpha`.

I'd love to fully deprecate the existing ColorPicker entirely, to be honest. Or export the new one under a separate name to allow for good tree-shaking (for example `SimpleColorPicker` or similar be a great name I think).

**Comment by ciampo** [comment]
> I think my proposal for the ColorPicker is to add a prop like `__unstableSimplified` which switches between the legacy and the new color picker, instead of trying to adapt the legacy props.
> 
> Primarily the issue I have the with the existing ColorPicker is that it's API is far too broad, especially in the change handlers.
> 
> It also uses "backwards" props like `disableAlpha` instead of positive boolean props like `enableAlpha`.
> 
> I'd love to fully deprecate the existing ColorPicker entirely, to be honest. Or export the new one under a separate name to allow for good tree-shaking (for example `SimpleColorPicker` or similar be a great name I think).

Thank you for elaborating. What you suggest is definitely a possibility. This kind of approach should be probably discussed in the context of coming up with a strategy that can work across the whole components library (for example, we may have a similar situation with the Flyout and the Slider component). Definitely not a blocker for this PR

**Comment by sarayourfriend** [comment]
Let's get #33784 and #33696 merged then revisit this.

**Comment by sarayourfriend** [comment]
@ciampo This is ready for another review when you have time. I think I'd like to merge this into the UI folder and then handle adapting/removing the existing color picker in favor of this one in a separate PR.

File: packages/components/src/select-control/styles/select-control-styles.ts [hunk 10]
```diff
@@ -29,7 +29,7 @@ const fontSizeStyles = ( { selectSize }: SelectProps ) => {
 		small: '11px',
 	};
 
-	const fontSize = sizes[ selectSize ];
+	const fontSize = sizes[ selectSize as Size ];
```
Comment (ciampo) [inline packages/components/src/select-control/styles/select-control-styles.ts:32]: Instead of typecasting, wouldn't it be better to add a check earlier in the function? Something like

```js
if ( ! selectSize ) return '';
```

File: packages/components/src/select-control/styles/select-control-styles.ts [hunk 11]
```diff
@@ -57,7 +57,7 @@ const sizeStyles = ( { selectSize }: SelectProps ) => {
 		},
 	};
 
-	const style = sizes[ selectSize ] || sizes.default;
+	const style = sizes[ selectSize as Size ] || sizes.default;
```
Comment (ciampo) [inline packages/components/src/select-control/styles/select-control-styles.ts:60]: Instead of typecasting, maybe we could use a ternary?

```suggestion
	const style = selectSize ? sizes[ selectSize ] : sizes.default;
```

File: packages/components/src/utils/hooks/use-controlled-value.ts [hunk 12]
```diff
@@ -27,8 +27,18 @@ export function useControlledValue< T >( {
...
+	if ( hasValue && typeof onChange === 'function' ) {
+		setValue = onChange;
+	} else if ( ! hasValue && typeof onChange === 'function' ) {
+		setValue = ( nextValue ) => {
+			onChange( nextValue );
+			setState( nextValue );
+		};
+	} else {
+		setValue = setState;
+	}
```
Comment (ciampo) [inline packages/components/src/utils/hooks/use-controlled-value.ts:41]: Does this "pattern" have a standard name (semi-controlled components)? It'd be nice to include it in future documentation with a brief explanation of its working principles (cc @diegohaz )

**Review by ciampo — COMMENTED** [review]
This is looking close! Here's some feedback, most of which is minor:

#### HSL constraints

When picking white as the color and switching to HLS mode, the H and S sliders are locked and can't be dragged, even though they don't appear disabled. In a similar fashion, the number input to the left of the slider can be changed only temporarily, since the new value will not be "committed" once the focus moves away from the input element.

https://user-images.githubusercontent.com/1083581/128700099-99349afb-877f-4c60-bb55-10aa21b578dd.mp4

While I understand the reason why these constraints are in place, I'm not sure this is the best UX possible — should we allow any value for H and S (within the correct range) instead?

Related to this, I noticed that, when the white color is selected, moving the dot across the square color palette doesn't cause a change in the H value. A side effect of this is that, when the white color is selected and a change in opacity happens, the H always gets reset to `0`.

https://user-images.githubusercontent.com/1083581/128701552-054c01a9-56d8-49dd-a26a-ceaf9bf4c360.mp4

#### Lack of unit tests

There are currently no unit tests — this is probably ok, given that the component is still experimental and lives in the UI folder. Although it would be better if we had some unit tests making sure that the component works as expected.

#### Folder structure

At the moment this component's folder structure is different from the usual one. Not a blocker for this PR, but we should probably look into "normalising" it once we reconcile this new Color Picker with the legacy one. 

#### Design

Most of the previous design-related feedback has been tackled (vertical spacing, letters and numbers vertical alignment, input and select controls).

We can have a final "design review" later / in a separate PR, definitely not a blocker here.

#### Pending feedback from previous review rounds

I can still see a few open conversations from the previous rounds: [1](https://github.com/WordPress/gutenberg/pull/33714#discussion_r679341598), [2](https://github.com/WordPress/gutenberg/pull/33714#discussion_r679369324), [3](https://github.com/WordPress/gutenberg/pull/33714#discussion_r679376032), [4](https://github.com/WordPress/gutenberg/pull/33714#discussion_r679377689), [5](https://github.com/WordPress/gutenberg/pull/33714#discussion_r679378418)

File: packages/components/src/utils/hooks/use-controlled-value.ts (hunk 12 above)
Comment (sarayourfriend) [reply packages/components/src/utils/hooks/use-controlled-value.ts:41]: It's not controlled unless you're passing a `value` prop, so semi-controlled seems misleading.

This just mirrors more closely how the intrinsic `input` controls work. I think it still falls under the controlled/uncontrolled dichotomy?

**Comment by sarayourfriend** [comment]
> HSL constraints

This is a side effect of always communicating in hex I guess. There's no way to encode these different values when L is at 100 in hex. We could switch the internals to always communicate in RGB or HSL for example, I think that would fix it 🤔 What do you think about that? It complicates the internals but fixes the UX you described.

**Comment by ciampo** [comment]
> We could switch the internals to always communicate in RGB or HSL for example, I think that would fix it 🤔 What do you think about that? It complicates the internals but fixes the UX you described.

I'd personally like to give it a shot if it's not too complicated. HSL seems like the format that would allow for the most flexibility — what do you think?

File: packages/components/src/utils/hooks/use-controlled-value.ts (hunk 12 above)
Comment (ciampo) [reply packages/components/src/utils/hooks/use-controlled-value.ts:41]: Gotcha. I think I'm not going to mention it necessarily in the guidelines, for now

**Comment by sarayourfriend** [comment]
I think we have to change the output/input value to also be HSL(A) otherwise the HSL encoding is immediately lost the second that it leaves the component and gets passed back in (in the case of a controlled component like in the stories).

**Comment by sarayourfriend** [comment]
I'm not sure how to solve this because changing the external contract to HSL(A) makes it a lot more complicated from a consumer perspective. But on the other hand trying to simplify it and accept _any_ color format as input, allows the user to incorrectly push a hex value which will cause the internals to no longer be able to encode valid HSL(A) values (even if they have no visual difference)

**Comment by ciampo** [comment]
Probably the best tradeoff would be to switch to HSL(A) for both the internal representation and for the accepted format of the `color` prop.

We could then expose the `colorize` function as a separate utility — useful in case the user of `ColorPicker` needs to convert color values to HSL(A).

**Comment by sarayourfriend** [comment]
> We could then expose the colorize function as a separate utility — useful in case the user of ColorPicker needs to convert color values to HSL(A).

I'd rather we didn't expose colorize directly... people can easily import tinycolor2 if they want rather than making it part of the public WP api.

File: packages/components/src/select-control/styles/select-control-styles.ts (hunk 10 above)
Comment (sarayourfriend) [reply packages/components/src/select-control/styles/select-control-styles.ts:32]: I was trying to avoid making any runtime changes to those components.

**Comment by sarayourfriend** [comment]
> At the moment this component's folder structure is different from the usual one. Not a blocker for this PR, but we should probably look into "normalising" it once we reconcile this new Color Picker with the legacy one.

What would you suggest? I'm mirroring the way lots of other of these high-level composition components are structured. I don't think an individual folder and README for each of these internal components makes any sense at all. Do you have a suggestion for this?

File: packages/components/src/select-control/styles/select-control-styles.ts (hunk 10 above)
Comment (ciampo) [reply packages/components/src/select-control/styles/select-control-styles.ts:32]: Makes sense. Feel free to keep it as-is in that case. We can refactor the code at a later point in a smaller, separate PR.

File: packages/components/src/select-control/styles/select-control-styles.ts (hunk 11 above)
Comment (ciampo) [reply packages/components/src/select-control/styles/select-control-styles.ts:60]: Same as https://github.com/WordPress/gutenberg/pull/33714#discussion_r685872996

File: packages/components/src/ui/color-picker/component.tsx [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+import { ColorInput } from './color-input';
+import { Picker } from './picker';
+import { useControlledValue } from '../../utils/hooks';
+
+import type { ColorType } from './types';
+
+interface ColorPickerProps {
+	enableAlpha?: boolean;
+	color?: ColorFormats.HSL | ColorFormats.HSLA;
+	onChange?: ( hexColor: ColorFormats.HSL | ColorFormats.HSLA ) => void;
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/component.tsx]: Nit

```suggestion
	onChange?: ( color: ColorFormats.HSL | ColorFormats.HSLA ) => void;
```

**Review by ciampo — APPROVED** [review]
After the switch to HSL(A)-based input, the UX has definitely improved!

I'm happy to work on unit tests, folder structure, and design refinements in follow-up PRs.

(cc @diegohaz in case you want to give this PR a final look)

> What would you suggest? I'm mirroring the way lots of other of these high-level composition components are structured. I don't think an individual folder and README for each of these internal components makes any sense at all. Do you have a suggestion for this?

I agree that one folder for each subcomponent doesn't make sense — especially because these subcomponents are not going to be exported.

My idea was more about splitting the component's "entry point" (e.g. `component.tsx`) from the remaining sub-components. 
Maybe having a few folders grouping sub-components with similar functionality, like `pickers/` (with `hex-color-picker` and `rgba-color-picker`) and `inputs/` (`color-input`, `hex-input`, `hsl-input`, `rgb-input`).

Another difference compared to "standard" g2 components is that there's no `hook.ts` file.

As explained before, these considerations should not be considered blockers for this PR.

**Comment by ciampo** [comment]
Actually, flagging these accessibility warnings from Storybook

Screenshot 2021-08-10 at 12 31 16

File: packages/components/src/ui/color-picker/component.tsx [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+						color={ safeColor }
+						colorType={ copyFormat || colorType }
+						enableAlpha={ enableAlpha }
+					/>
+				) }
+				<Button
+					onClick={ () => setShowInputs( ! showInputs ) }
+					icon={ moreVertical }
+					isPressed={ showInputs }
+				/>
```
Comment (diegohaz) [inline packages/components/src/ui/color-picker/component.tsx]: We should pass a `label` prop to this button.

**Comment by diegohaz** [comment]
> Also flagging these accessibility warnings from Storybook

The first one seems to be a problem with the react-colorful package. The `aria-valuenow` attribute is required on `role="slider"` elements: https://w3c.github.io/aria/#slider

Do we have control over the elements that react-colorful renders? If so, we should be able to add that prop.

**Comment by sarayourfriend** [comment]
> Do we have control over the elements that react-colorful renders

We don't 😞 

It looks like they've already done _some_ accessibility work so I'm sure they'd be open to a PR to fix this: https://github.com/omgovich/react-colorful/issues/57

I'll see about opening one upstream.

**Comment by sarayourfriend** [comment]
> Another difference compared to "standard" g2 components is that there's no hook.ts file

Does it need a hook?

It feels like we shouldn't thoughtlessly follow the hook + component pattern if it just adds redirection/complexity to something that can just be a regular component (especially if the hook isn't meant to be reusable, which is the primary benefit of extrapolating things into hooks, but in this case what would be the reusable version of the hook for this?)

**Review by ciampo — APPROVED** [review]
All accessibility issues that can be fixed on our end are now solved.

> I'll see about opening one upstream.

That sounds great, thank you!

> Does it need a hook?

You actually make a good point about the logic of this component not needing to be "reusable". I agree with you, it doesn't need a hook.

**Comment by diegohaz** [comment]
Great work! 🎉

For the record, here's the issue about the lack of the `aria-valuenow` attribute: https://github.com/omgovich/react-colorful/issues/152

**Comment by omgovich** [comment]
Hi guys! Thanks for choosing react-colorful for your project!
I will try to improve a11y as soon as possible 🤝

P.S. I noticed that your project using `tinycolor2` which is pretty slow, heavy, and requires a separate `@types/tinycolor2` package to make it work in TS project. So you might be interested in using `colord` which is 3 times lighter, 3 times faster, has all of the types, modular and extendable.
https://colord.omgovich.ru/ 
https://github.com/omgovich/colord

The API is kinda the same as `tinycolor2` has, so the migration shouldn't be a big deal if you decide to do that).

For example, `cssnano` recently migrated from `color` (kinda the same old-school package as `tinycolor2`) to `colord` and after that their performance became 2x faster.

File: packages/components/src/ui/color-picker/color-display.tsx [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+				default:
+				case 'hex': {
+					const colorized = colorize( color );
+					return enableAlpha
+						? colorized.toHex8String()
+						: colorized.toHexString();
+				}
+			}
+		},
+		() => setCopiedColor( colorize( color ).toHex8String() )
```
Comment (ciampo) [inline packages/components/src/ui/color-picker/color-display.tsx:137]: I just noticed a minor UX improvement that we could make:

1. User clicks on the color, value is copied in the clipboard. The tooltip changes from "Copy" to "Copied!"
2. User interacts with other parts of the UI, maybe copies another string of text
3. User remembers that they need the color value, and so they hover over the color string again. But the tooltip keeps saying "Copied!", instead of "Copy"

This to say, should we use a timer to unset the `copiedColor` after some time has passed?
